Backport Structured Audit Logging

albertzaharovits opened the pull request:
This is the backport of https://github.com/elastic/elasticsearch/pull/31931 .

Following the conclusion on https://github.com/elastic/elasticsearch/issues/31046#issuecomment-414366582 , this backport resurrects the previous logging format, such that **both** are enabled side by side for the 6.x releases. This is achieved by adding another appender with it's own pattern layout in the log4j configuration. This pattern layout "emulates" the previous format. Codewise, the same log4j `StringMapMessage` event is used to generate the log lines in both formats. There are truly no guarantees that the "emulated" format is 100% the old one (besides my 🦅  eyes).

 Unit tests check **both** layouts by parsing the `log4j2.properties` file as a resource. There are some qa integ tests in the sql projects that parse the audit log when security is enabled. Those only parse the new **format**.

This commit https://github.com/elastic/elasticsearch/pull/33894/commits/1ca9d19c1e7ebd8b25ae1f535c299a429336bdd9 adds the "emulated" format (the other commit is the backport).

elasticmachine commented:
Pinging @elastic/es-security

albertzaharovits commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/audit/logfile/CapturingLogger.java:
```diff
@@ -8,45 +8,134 @@
+            }
+        }
+        return true;
+    }
```
Is empty can optionally check multiple appenders.

jaymode commented:
> There are truly no guarantees that the "emulated" format is 100% the old one (besides my 🦅 eyes).

I am having a hard time with this. I think we should put in more effort here to validate or keep the existing way as is; I think we can accomplish this by having a special logger for the legacy format. This allows us to keep the pre-existing log statements alongside the new map format that uses a different logger. What do you think?

albertzaharovits commented:
> I think we can accomplish this by having a special logger for the legacy format.

@jaymode I am thinking of cherrie-picking the old `LoggingAuditTrail`, rename it, and then enabling it when `logfile` output is on. This is the only way, in my opinion, to guarantee that the legacy format is 100% identical. Copy pasting the message building logic alongside the new `StringMapMessage` building is no better that what I have proposed herein.
I remember previously considering this, but don't remember the conclusion (probably something about the tests), I will give it a try, though.

jaymode commented:
I am good with that approach @albertzaharovits.

albertzaharovits commented:
@jaymode  I have reverted the "emulated" format logic and cherry-picked `LoggingAuditTrail` and `LoggingAuditTrailTests` from 6.x and renamed them `DeprecatedLoggingAuditTrail` and `DeprecatedLoggingAuditTrailTests` respectively. 
The `DeprecatedLoggingAuditTrail` had all it's settings, as well as `EventFilterPolicyRegistry`, `EventFilterPolicy` and `AuditEventMetaInfo` static inner classes removed, as it is using the ones from the `LoggingAuditTrail`.

albertzaharovits commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/audit/logfile/CapturingLogger.java:
```diff
@@ -52,42 +54,88 @@ public static Logger newCapturingLogger(final Level level, @Nullable StringLayou
 
     /**
-     * Checks if the logger's appender has captured any events.
+     * Checks if the logger's appender(s) has captured any events.
```
this is not relevant anymore, the commit has been reverted.

albertzaharovits commented on x-pack/plugin/core/src/main/config/log4j2.properties:
```diff
@@ -1,9 +1,101 @@
+appender.deprecated_audit_rolling.type = RollingFile
+appender.deprecated_audit_rolling.name = deprecated_audit_rolling
+appender.deprecated_audit_rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_access.log
```
@bizybot See the discussion starting here https://github.com/elastic/elasticsearch/issues/31046#issuecomment-413783224

Basically, Filebeat has to support whatever format we're defaulting too. So if we're not enabling the new appender in 6.x Filebeat will not use this in 6.x, so we might as well not port this to 6.x because Filebeat is the important reason we do this (sort of, the root reason is to get rid of the `IndexAuditTrail`, but we need Filebeat to consume or logs before we can do that).

And since we ❤️ our users, we don't want to break any systems they have set up right now that are consuming the present format. So, the compromise is to enable both of them.

albertzaharovits commented on x-pack/plugin/core/src/main/config/log4j2.properties:
```diff
@@ -1,9 +1,101 @@
+appender.deprecated_audit_rolling.policies.time.interval = 1
+appender.deprecated_audit_rolling.policies.time.modulate = true
+
 appender.audit_rolling.type = RollingFile
```
This is a good idea!
I have raised https://github.com/elastic/elasticsearch/issues/34321 to discuss this further . 
I am concerned that we might not be able to decide this for every use case for every OS , so
your suggestion might end up in the docs rather then the default distribution.
But, let's discuss this separately, and keep this PR concerning the format only.

albertzaharovits commented:
@bizybot @jaymode this is ready for another round. Would you please take another look?

albertzaharovits commented:
test this please

albertzaharovits commented:
`org.gradle.launcher.daemon.client.DaemonDisappearedException: Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)` run gradle build tests

albertzaharovits commented:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/1647/consoleFull
Failed with:
```
...
17:22:43     Caused by: java.io.FileNotFoundException: /tmp/.gradle-test-kit-jenkins/caches/4.10/scripts-remapped/build_5ja2t0gai22z6i9he69emifxd/bierx9xgpipif9vv3my944rer/cp_proj525a2f1efef12369079491b50deb9f6c/metadata/metadata.bin (No such file or directory)
17:22:43     	at org.gradle.groovy.scripts.internal.DefaultScriptCompilationHandler.loadFromDir(DefaultScriptCompilationHandler.java:209)
17:22:43     	... 106 more
...
17:22:43     > Configure project :
17:22:43     Evaluating root project 'elasticsearch-build-resources' using build file '/var/lib/jenkins/workspace/elastic+elasticsearch+pull-request/buildSrc/src/testKit/elasticsearch-build-resources/build.gradle'.
17:22:43 
17:22:43     FAILURE: Build failed with an exception.
17:22:43 
17:22:43     * What went wrong:
17:22:43     A problem occurred configuring root project 'elasticsearch-build-resources'.
17:22:43     > Failed to deserialize script metadata extracted for build file '/var/lib/jenkins/workspace/elastic+elasticsearch+pull-request/buildSrc/src/testKit/elasticsearch-build-resources/build.gradle'
```
run gradle build tests

jaymode commented on x-pack/plugin/security/build.gradle:
```diff
@@ -146,6 +146,7 @@ processResources {
 
 sourceSets.test.resources {
     srcDir '../core/src/test/resources'
+    srcDir '../core/src/main/config'
```
can you add a comment about why we pull this in for log4j config

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:
```diff
@@ -422,6 +423,9 @@ protected XPackLicenseState getLicenseState() {
                         auditTrails.add(new LoggingAuditTrail(settings, clusterService, threadPool));
+                        // also enabling the deprecated format. To disable it, remove it's associated
+                        // appender in the log4j2.properties file
+                        auditTrails.add(new DeprecatedLoggingAuditTrail(settings, clusterService, threadPool));
```
we should have documentation about how to disable this. I am fine if you want to handle this separately, but please open up a issue if that is what you choose to do

albertzaharovits commented:
```
10:59:50 FAILURE: Build failed with an exception.
10:59:50 
10:59:50 * What went wrong:
10:59:50 Could not set the value of environment variable 'ghprbPullLongDescription': could not convert string to current locale
10:59:50 
10:59:50 * Try:
10:59:50 Run with --debug option to get more log output. Run with --scan to get full insights.
10:59:50 
10:59:50 * Exception is:
10:59:50 net.rubygrapefruit.platform.NativeException: Could not set the value of environment variable 'ghprbPullLongDescription': could not convert string to current locale
10:59:50 	at net.rubygrapefruit.platform.internal.DefaultProcess.setEnvironmentVariable(DefaultProcess.java:71)
10:59:50 	at net.rubygrapefruit.platform.internal.WrapperProcess.setEnvironmentVariable(WrapperProcess.java:84)
10:59:50 	at org.gradle.internal.nativeintegration.processenvironment.NativePlatformBackedProcessEnvironment.setNativeEnvironmentVariable(NativePlatformBackedProcessEnvironment.java:36)
10:59:50 	at org.gradle.internal.nativeintegration.processenvironment.AbstractProcessEnvironment.setEnvironmentVariable(AbstractProcessEnvironment.java:67)
10:59:50 	at org.gradle.internal.nativeintegration.processenvironment.AbstractProcessEnvironment.maybeSetEnvironment(AbstractProcessEnvironment.java:41)
10:59:50 	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:65)
10:59:50 	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
10:59:50 	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
10:59:50 	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
10:59:50 	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:295)
10:59:50 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
10:59:50 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
10:59:50 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
10:59:50 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
10:59:50 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
10:59:50 	at java.lang.Thread.run(Thread.java:748)
```
run gradle build tests

ruflin commented:
@ycombinator FYI, this was backported.
